xlive/XLive/UserManagement/CUser.cpp

```diff
@@ -98,51 +112,41 @@ void CUserManagement::CreateUser(const XNADDR* pxna, BOOL user)
 
 		this->secure_map[hostpair] = secure;
 		this->secure_map[hostpair_1000] = secure;
-		this->cusers[secure] = nUser;
-		this->xnmap[secure] = pxna->ina.s_addr;
 
 		this->pmap_a[secure] = nPort_base;
 		this->pmap_b[secure] = nPort_join;
 	}
-	else
-	{
-		/*
-			If the user variable is true, we update cusers "array" to map the secure address to the newly created object...
-			Then map the secure address to the WAN IP (XN->ina) using xnmap, where secure address is the key to the array.
-			Then map the abenet to the secure address via xntosecure[ab], where AB is the key to the array.
-		*/
-		this->cusers[secure] = nUser;
-		this->xnmap[secure] = pxna->ina.s_addr;
-	}
-
-	nUser->bValid = true;
+	this->xnmap[secure] = pxna->ina.s_addr;
 }
 
-// FIXME
 void CUserManagement::UnregisterSecureAddr(const IN_ADDR ina)
 {
 	CUser* to_remove_user = cusers[ina.s_addr];
 	if (to_remove_user != nullptr)
 	{
-		std::pair<ULONG, short> smap, smap1;
-		if (to_remove_user->xnaddr.ina.s_addr == H2Config_ip_wan)
-		{
-			smap = std::make_pair(H2Config_ip_lan, to_remove_user->xnaddr.wPortOnline);
-			smap1 = std::make_pair(H2Config_ip_lan, ntohs(htons(to_remove_user->xnaddr.wPortOnline) + 1));
-		}
-		else
+		if (xnmap[ina.s_addr])
+			xnmap.erase(ina.s_addr);
+
+		if (pmap_a[ina.s_addr])
+			pmap_a.erase(ina.s_addr);
+
+		if (pmap_b[ina.s_addr])
+			pmap_b.erase(ina.s_addr);
+
+		if (cusers[ina.s_addr])
+			cusers.erase(ina.s_addr);
+
+		for (auto it : secure_map)
 		{
-			smap = std::make_pair(to_remove_user->xnaddr.ina.s_addr, to_remove_user->xnaddr.wPortOnline);
-			smap1 = std::make_pair(to_remove_user->xnaddr.ina.s_addr, ntohs(htons(to_remove_user->xnaddr.wPortOnline) + 1));
+			if (it.second == ina.s_addr) {
+				secure_map.erase(it.first);
```

Comment:
not sure if it's safe to remove elements while iterating? Could crash or skip an element... Might be better to use std::remove_if

Reply:
it exits the loop after it finds the element, shouldn't cause any issues

Reply:
ok yeah that's fine then.

Reply:
is it fine though? how many different threads might access this map at the same time? if we can guarantee single threaded access to the map then it seems okay, but in reality it should be thread safe